pkg/ingester/streams_map.go

```diff
@@ -72,32 +101,49 @@ func (m *streamsMap) load(mp *sync.Map, key interface{}) (*stream, bool) {
 	return nil, false
 }
 
-func (m *streamsMap) loadOrStoreNew(mp *sync.Map, key interface{}, newStreamFn func() (*stream, error)) (*stream, bool, error) {
-	s, ok := m.load(mp, key)
-
-	if ok {
-		return s, true, nil
-	}
-
-	m.consistencyMtx.Lock()
-	defer m.consistencyMtx.Unlock()
-	// Double check
-	s, ok = m.load(mp, key)
-
-	if ok {
-		return s, true, nil
-	}
-
-	s, err := newStreamFn()
-	if err != nil {
-		return nil, false, err
-	}
+func (m *streamsMap) store(key interface{}, s *stream) {
 	if labelsString, ok := key.(string); ok {
 		m.streams.Store(labelsString, s)
 	} else {
 		m.streams.Store(s.labelsString, s)
 	}
 	m.streamsByFP.Store(s.fp, s)
 	m.streamsCounter.Inc()
-	return s, false, nil
+}
+
+// newStreamFn: Called if not loaded, with consistencyMtx locked. Must not be nil
+// postLoadFn: Called if loaded, with consistencyMtx read-locked at least. Can be nil
+func (m *streamsMap) loadOrStoreNew(mp *sync.Map, key interface{}, newStreamFn func() (*stream, error), postLoadFn func(*stream) error) (*stream, bool, error) {
+	var s *stream
+	var loaded bool
+	var err error
+	m.WithRLock(func() {
+		if s, loaded = m.load(mp, key); loaded {
+			if postLoadFn != nil {
+				err = postLoadFn(s)
+			}
+		}
+	})
+
+	if loaded {
+		return s, true, err
+	}
```

Comment:
I side with Callum here. This is easier to comprehend IMHO
```suggestion
m.consistencyMtx.RLock()
if s, loaded = m.load(mp, key); loaded {
    if postLoadFn != nil {
        err = postLoadFn(s)
	}
	m.consistencyMtx.RUnlock()
	return s, true, err
}
m.consistencyMtx.RUnlock()
```